crates/phactory/src/system/mod.rs

```diff
@@ -866,6 +815,81 @@ impl<Platform: pal::Platform> System<Platform> {
             info!("Gatekeeper: successfully decrypt received master key");
             self.set_master_key(master_pair, true);
         }
+        Ok(())
+    }
+
+    fn process_contract_key_distribution(
+        &mut self,
+        block: &mut BlockInfo,
+        origin: MessageOrigin,
+        event: DispatchContractKeyEvent<chain::Hash, chain::BlockNumber, chain::AccountId>,
+    ) -> Result<(), TransactionError> {
+        if !origin.is_gatekeeper() {
+            error!("Invalid origin {:?} sent a {:?}", origin, event);
+            return Err(TransactionError::BadOrigin);
+        }
+
+        // TODO(shelven): forget contract key after expiration time
+        let keypair = sr25519::Pair::restore_from_secret_key(&event.secret_key);
+        let contract_key = ContractKey(keypair);
+        let contract_pubkey = contract_key.public();
+        let contract_info = event.contract_info;
+        match contract_info.code_index {
+            CodeIndex::NativeCode(_contract_id) => {
+                // TODO(shelven): launch native code instance
+            }
+            CodeIndex::WasmCode(code_hash) => {
+                let code = chain_state::read_contract_code(block.storage, code_hash);
+                if let Some(code) = code {
+                    if !self.contract_keys.contains_key(&contract_pubkey) {
+                        self.contract_keys
+                            .insert(contract_pubkey, contract_key.clone());
+
+                        let group_id = chain::Hash::from_low_u64_be(contract_info.group_id);
+                        let result = self.contract_groups.instantiate_contract(
+                            group_id,
+                            contract_info.deployer.clone(),
+                            code,
+                            contract_info.instantiate_data,
+                            contract_info.salt,
+                            &contract_key,
+                            block.block_number,
+                            block.now_ms,
+                        );
+                        match result {
+                            Err(err) => {
+                                error!(
```

Comment:
Do you want to continue to the logic below in this case?

Reply:
done